taichi/ir/frontend_ir.cpp

```diff
@@ -180,7 +180,8 @@ void BinaryOpExpression::type_check(CompileConfig *config) {
                     binary_op_type_symbol(type), lhs->ret_type->to_string(),
                     rhs->ret_type->to_string()));
   };
-  if (!lhs_type->is<PrimitiveType>() || !rhs_type->is<PrimitiveType>())
+  if (!(lhs_type->is<PrimitiveType>() && rhs_type->is<PrimitiveType>()) &&
+      !(lhs_type->is<TensorType>() && rhs_type->is<TensorType>()))
```

Comment:
I was also thinking of this check when implementing matrix-matrix operations. This will disallow operations like `mat + 1`, which is a broadcast addition. (but the workaround is to convert the constant 1 to a TensorType MatrixExpression before going through `type_check`.

Reply:
Good point, thanks! Let me add a `TI_NOT_IMPLEMENTED` for the case where broadcasting is needed.

Reply:
I was thinking we should perform broadcasting along with type promotion